mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,761 @@
+static uint64_t getParallelismCost(ArrayRef<unsigned> perm,
+                                   ArrayRef<bool> loopCarriedDV,
+                                   ArrayRef<unsigned> iterCounts) {
+  uint64_t totalcost = 0;
+  uint64_t thisLoopcost = 1;
+  for (unsigned i = 0; i < perm.size(); i++) {
+    if (!loopCarriedDV[perm[i]])
+      continue;
+    thisLoopcost = 1;
+    for (unsigned j = i + 1; j < perm.size(); j++)
+      thisLoopcost *= iterCounts[perm[j]];
+    totalcost += thisLoopcost;
+  }
+  return totalcost;
+}
+
+/// Calculates a representative temporal reuse cost for a given permutation of
+/// the loop nest. A lower value returned means higher temporal reuse.
+static uint64_t getTemporalReuseCost(
+    ArrayRef<unsigned> permutation, ArrayRef<unsigned> loopIterationCounts,
+    DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+        &loopAccessMatrices,uint64_t sentinel) {
+
+  // Initially, we assume no temporal reuse. The cost is a big value - sentinel.
+  uint64_t cost = sentinel;
+
+  // Start with the innermost loop to check if the access matrix for an op has
+  // all zeros in the respective column. If yes, there is a O(n) reuse. The
+  // reuse gets multiplied for each loop index until the first loop index with
+  // no reuse is encountered.
+  uint64_t temporalReuse = 1;
+  for (auto &accessMatrixOpPair : loopAccessMatrices) {
+    temporalReuse = 1;
+    SmallVector<SmallVector<int64_t, 4>, 4> accessMatrix =
+        accessMatrixOpPair.second;
+    for (int i = permutation.size() - 1; i >= 0; i--) {
+      bool isColumnAllZeros = true;
+      for (SmallVector<int64_t, 4> &row : accessMatrix) {
+        if (row[permutation[i]] != 0) {
+          isColumnAllZeros = false;
+          break;
+        }
+      }
+      if (!isColumnAllZeros)
+        break;
+      temporalReuse *= loopIterationCounts[permutation[i]];
+    }
+    // Increase in temporalReuse decreases the cost.
+    cost -= temporalReuse;
+  }
+  return cost;
+}
+
+/// Removes `dstOp` from its current group and inserts it into the `srcOp`'s
+/// group. Updates `groupId` to reflect the changes.
+static void insertIntoReferenceGroup(
+    Operation *srcOp, Operation* dstOp, DenseMap<Operation*, unsigned> &groupId,
+    SmallVector<llvm::SmallSet<Operation *, 8>, 8> &referenceGroups) {
+  referenceGroups[groupId[srcOp]].insert(
+      referenceGroups[groupId[dstOp]].begin(),
+      referenceGroups[groupId[dstOp]].end());
+  referenceGroups.erase(referenceGroups.begin() + groupId[dstOp]);
+  // Insert operation results in same group-id for both instructions.
+  groupId[dstOp] = groupId[srcOp];
+}
+
+/// Groups ops in `loadAndStoreOps` into `referenceGroups` based on whether or
+/// not they exhibit group-temporal or group-spatial reuse with respect to an
+/// affine.for op present at depth `innermostIndex` in the original loop nest.
+///
+/// Please refer Steve Carr et. al for a detailed description.
```

Comment:
refer to

Reply:
Fixed. Here and everywhere else. Thanks.